docs/proposals/20230726-integrate-keda-http-add-on.md

```diff
@@ -0,0 +1,249 @@
+- **Auto-scaling**:
+  1. In step 3 of the Request process, the Interceptor Deployment counts the requests, which are periodically collected and aggregated by the External Scaler Deployment.
+  2. External Scaler Deployment reports request counts to the KEDA operator.
+  3. KEDA scales the target (Function) that reaches the threshold value according to the scaling options.
+- **Inject Dapr sidecar**:
+  1. Dapr control plane includes the [dapr-sidecar-injector service](https://docs.dapr.io/operations/hosting/kubernetes/kubernetes-overview/), which watches for new pods with the `dapr.io/enabled` annotation and injects a container with the `daprd` process within the pod.
+  2. In this case, sidecar arguments can be passed through annotations as outlined in the **Kubernetes annotations** column in [this table](https://docs.dapr.io/reference/arguments-annotations-overview/).
+
+![sync runtime overview](../images/integrate_keda_http-add-on.png)
+
+![KEDA+Dapr architecture overview](../images/function-keda-dapr.png)
+
+
+
+### CRD
+
+Add an `HTTPScaledObject ` type in `OpenFunction/apis/core/v1beta2/serving_types.go`:
+
+```go
+type HTTPScaledObject struct {
+	// Target metric value
+	// +optional
+	TargetPendingRequests *int32 `json:"targetPendingRequests,omitempty"`
+	// Cooldown period value
+	// +optional
+	CooldownPeriod *int32 `json:"cooldownPeriod,omitempty"`
+}
+
+type KedaScaleOptions struct {
+	// +optional
+	ScaledObject *KedaScaledObject `json:"scaledObject,omitempty"`
+	// +optional
+	HTTPScaledObject *HTTPScaledObject `json:"httpScaledObject,omitempty"`
+	// +optional
+	ScaledJob *KedaScaledJob `json:"scaledJob,omitempty"`
+	// Triggers are used to specify the trigger sources of the function.
+	// The Keda (ScaledObject, ScaledJob) configuration in ScaleOptions cannot take effect without Triggers being set.
+	// +optional
+	Triggers []kedav1alpha1.ScaleTriggers `json:"triggers,omitempty"`
+}
+```
+
+To differ `keda-http addon` runtime from `knative` runtime, we can add an `Engine` section into `HttpTrigger`:
+
+```go
+type Engine string
+
+const (
+	HookPolicyAppend   = "Append"
+	HookPolicyOverride = "Override"
+
+	WorkloadTypeJob                = "Job"
+	WorkloadTypeStatefulSet        = "StatefulSet"
+	WorkloadTypeDeployment         = "Deployment"
+	HttpRuntimeKnative      Engine = "knative"
+	HttpRuntimeKeda         Engine = "keda"
+)
+
+type HttpTrigger struct {
+	// The port on which the function will be invoked
+	Port *int32 `json:"port,omitempty"`
+	// Information needed to make HTTPRoute.
+	// Will attempt to make HTTPRoute using the default Gateway resource if Route is nil.
+	//
+	// +optional
+	Route *RouteImpl `json:"route,omitempty"`
+	// Http function runtime engine, can be set to knative or keda, default to knative if not set
+	// +optional
+	Engine *Engine `json:"engine,omitempty"`
```

Comment:
@wanjunlei @wrongerror @tpiperatgod are you ok with this field or any other suggestion？

Reply:
lgtm